File: app/src/main/java/me/zhanghai/android/files/filejob/FileJobActionDialogFragment.java
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filejob;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.CheckBox;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatDialogFragment;
import androidx.core.widget.NestedScrollView;
import butterknife.BindView;
import butterknife.ButterKnife;
import me.zhanghai.android.files.R;
import me.zhanghai.android.files.compat.AlertDialogBuilderCompat;
import me.zhanghai.android.files.util.BundleBuilder;
import me.zhanghai.android.files.util.RemoteCallback;
import me.zhanghai.android.files.util.ViewUtils;

public class FileJobActionDialogFragment extends AppCompatDialogFragment {

    private static final String KEY_PREFIX = FileJobActionDialogFragment.class.getName() + '.';

    private static final String EXTRA_TITLE = KEY_PREFIX + "TITLE";
    private static final String EXTRA_MESSAGE = KEY_PREFIX + "MESSAGE";
    private static final String EXTRA_SHOW_ALL = KEY_PREFIX + "SHOW_ALL";
    private static final String EXTRA_POSITIVE_BUTTON_TEXT = KEY_PREFIX + "POSITIVE_BUTTON_TEXT";
    private static final String EXTRA_NEGATIVE_BUTTON_TEXT = KEY_PREFIX + "NEGATIVE_BUTTON_TEXT";
    private static final String EXTRA_NEUTRAL_BUTTON_TEXT = KEY_PREFIX + "NEUTRAL_BUTTON_TEXT";
    private static final String EXTRA_LISTENER = KEY_PREFIX + "LISTENER";

    private static final String STATE_ALL_CHECKED = KEY_PREFIX + "ALL_CHECKED";

    @NonNull
    private CharSequence mTitle;
    @NonNull
    private CharSequence mMessage;
    private boolean mShowAll;
    @Nullable
    private CharSequence mPositiveButtonText;
    @Nullable
    private CharSequence mNegativeButtonText;
    @Nullable
    private CharSequence mNeutralButtonText;
    @Nullable
    private RemoteCallback mListener;

    @Nullable
    private View mView;

    @BindView(R.id.all)
    CheckBox mAllCheck;

    public static void putArguments(@NonNull Intent intent, @NonNull CharSequence title,
                                    @NonNull CharSequence message, boolean showAll,
                                    @Nullable CharSequence positiveButtonText,
                                    @Nullable CharSequence negativeButtonText,
                                    @Nullable CharSequence neutralButtonText,
                                    @NonNull Listener listener) {
        intent
                .putExtra(EXTRA_TITLE, title)
                .putExtra(EXTRA_MESSAGE, message)
                .putExtra(EXTRA_SHOW_ALL, showAll)
                .putExtra(EXTRA_POSITIVE_BUTTON_TEXT, positiveButtonText)
                .putExtra(EXTRA_NEGATIVE_BUTTON_TEXT, negativeButtonText)
                .putExtra(EXTRA_NEUTRAL_BUTTON_TEXT, neutralButtonText)
                .putExtra(EXTRA_LISTENER, new RemoteCallback(new ListenerAdapter(listener)));
    }

    @NonNull
    public static FileJobActionDialogFragment newInstance(@NonNull Intent intent) {
        //noinspection deprecation
        FileJobActionDialogFragment fragment = new FileJobActionDialogFragment();
        fragment.setArguments(intent.getExtras());
        return fragment;
    }

    /**
     * @deprecated Use {@link #newInstance(Intent)} instead.
     */
    public FileJobActionDialogFragment() {}

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle arguments = getArguments();
        mTitle = arguments.getCharSequence(EXTRA_TITLE);
        mMessage = arguments.getCharSequence(EXTRA_MESSAGE);
        mShowAll = arguments.getBoolean(EXTRA_SHOW_ALL);
        mPositiveButtonText = arguments.getCharSequence(EXTRA_POSITIVE_BUTTON_TEXT);
        mNegativeButtonText = arguments.getCharSequence(EXTRA_NEGATIVE_BUTTON_TEXT);
        mNeutralButtonText = arguments.getCharSequence(EXTRA_NEUTRAL_BUTTON_TEXT);
        mListener = arguments.getParcelable(EXTRA_LISTENER);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);

        if (mShowAll) {
            outState.putBoolean(STATE_ALL_CHECKED, mAllCheck.isChecked());
        }
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {

        Context context = requireContext();
        int theme = getTheme();
        if (mShowAll) {
            mView = ViewUtils.inflateWithTheme(R.layout.file_job_action_dialog_view, context,
                    theme);
            ButterKnife.bind(this, mView);
            if (savedInstanceState != null) {
                mAllCheck.setChecked(savedInstanceState.getBoolean(STATE_ALL_CHECKED));
            }
        }

        AlertDialog dialog = AlertDialogBuilderCompat.create(context, theme)
                .setTitle(mTitle)
                .setMessage(mMessage)
                .setPositiveButton(mPositiveButtonText, this::onDialogButtonClick)
                .setNegativeButton(mNegativeButtonText, this::onDialogButtonClick)
                .setNeutralButton(mNeutralButtonText, this::onDialogButtonClick)
                .create();
        dialog.setCanceledOnTouchOutside(false);
        return dialog;
    }

    private void onDialogButtonClick(@NonNull DialogInterface dialog, int which) {
        Action action;
        switch (which) {
            case DialogInterface.BUTTON_POSITIVE:
                action = Action.POSITIVE;
                break;
            case DialogInterface.BUTTON_NEGATIVE:
                action = Action.NEGATIVE;
                break;
            case DialogInterface.BUTTON_NEUTRAL:
                action = Action.NEUTRAL;
                break;
            default:
                throw new IllegalArgumentException(Integer.toString(which));
        }
        sendResult(action, mShowAll && mAllCheck.isChecked());
        requireActivity().finish();
    }

    @Override
    public void onStart() {
        super.onStart();

        if (mView != null) {
            AlertDialog dialog = (AlertDialog) requireDialog();
            NestedScrollView scrollView = dialog.findViewById(R.id.scrollView);
            LinearLayout linearLayout = (LinearLayout) scrollView.getChildAt(0);
            linearLayout.addView(mView);
            mView = null;
        }
    }

    @Override
    public void onCancel(@NonNull DialogInterface dialog) {
        super.onCancel(dialog);

        sendResult(Action.CANCELED, false);
        requireActivity().finish();
    }

    public void onFinish() {
        sendResult(Action.CANCELED, false);
    }

    private void sendResult(@NonNull Action action, boolean all) {
        if (mListener != null) {
            ListenerAdapter.sendResult(mListener, action, all);
            mListener = null;
        }
    }

    public enum Action {
        POSITIVE,
        NEGATIVE,
        NEUTRAL,
        CANCELED
    }

    public interface Listener {

        void onAction(@NonNull Action action, boolean all);
    }

    private static class ListenerAdapter implements RemoteCallback.Listener {

        private static final String KEY_PREFIX = Listener.class.getName() + '.';

        private static final String RESULT_ACTION = KEY_PREFIX + "ACTION";
        private static final String RESULT_ALL = KEY_PREFIX + "ALL";

        @NonNull
        private final Listener mListener;

        public ListenerAdapter(@NonNull Listener listener) {
            mListener = listener;
        }

        public static void sendResult(@NonNull RemoteCallback listener, @NonNull Action action,
                                      boolean all) {
            listener.sendResult(new BundleBuilder()
                    .putSerializable(RESULT_ACTION, action)
                    .putBoolean(RESULT_ALL, all)
                    .build());
        }

        @Override
        public void onResult(Bundle result) {
            Action action = (Action) result.getSerializable(RESULT_ACTION);
            boolean all = result.getBoolean(RESULT_ALL);
            mListener.onAction(action, all);
        }
    }
}

File: app/src/main/java/me/zhanghai/android/files/filejob/FileJobActionDialogFragment.kt
/*
 * Copyright (c) 2019 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filejob

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.os.Parcel
import android.widget.LinearLayout
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.core.view.isVisible
import androidx.core.widget.NestedScrollView
import androidx.lifecycle.lifecycleScope
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.launch
import kotlinx.parcelize.Parceler
import kotlinx.parcelize.Parcelize
import kotlinx.parcelize.WriteWith
import me.zhanghai.android.files.R
import me.zhanghai.android.files.compat.requireViewByIdCompat
import me.zhanghai.android.files.databinding.FileJobActionDialogViewBinding
import me.zhanghai.android.files.provider.common.PosixFileStore
import me.zhanghai.android.files.util.ActionState
import me.zhanghai.android.files.util.ParcelableArgs
import me.zhanghai.android.files.util.ParcelableParceler
import me.zhanghai.android.files.util.ParcelableState
import me.zhanghai.android.files.util.RemoteCallback
import me.zhanghai.android.files.util.args
import me.zhanghai.android.files.util.finish
import me.zhanghai.android.files.util.getArgs
import me.zhanghai.android.files.util.getState
import me.zhanghai.android.files.util.isReady
import me.zhanghai.android.files.util.isRunning
import me.zhanghai.android.files.util.layoutInflater
import me.zhanghai.android.files.util.putArgs
import me.zhanghai.android.files.util.putState
import me.zhanghai.android.files.util.readParcelable
import me.zhanghai.android.files.util.showToast
import me.zhanghai.android.files.util.viewModels

class FileJobActionDialogFragment : AppCompatDialogFragment() {
    private val args by args<Args>()

    private val viewModel by viewModels { { FileJobActionViewModel() } }

    private lateinit var binding: FileJobActionDialogViewBinding

    private var isListenerNotified = false

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putState(State(binding.allCheck.isChecked))
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog =
        MaterialAlertDialogBuilder(requireContext(), theme)
            .setTitle(args.title)
            .setMessage(args.message)
            .apply {
                binding = FileJobActionDialogViewBinding.inflate(context.layoutInflater)
                val hasReadOnlyFileStore = args.readOnlyFileStore != null
                binding.remountButton.isVisible = hasReadOnlyFileStore
                if (hasReadOnlyFileStore) {
                    updateRemountButton()
                    binding.remountButton.setOnClickListener { remount() }
                }
                binding.allSpace.isVisible = !hasReadOnlyFileStore && args.showAll
                binding.allCheck.isVisible = args.showAll
                if (savedInstanceState != null) {
                    binding.allCheck.isChecked = savedInstanceState.getState<State>().isAllChecked
                }

                if (hasReadOnlyFileStore) {
                    lifecycleScope.launchWhenStarted {
                        launch { viewModel.remountState.collect { onRemountStateChanged(it) } }
                    }
                }
            }
            .setPositiveButton(args.positiveButtonText, ::onDialogButtonClick)
            .setNegativeButton(args.negativeButtonText, ::onDialogButtonClick)
            .setNeutralButton(args.neutralButtonText, ::onDialogButtonClick)
            .create()
            .apply { setCanceledOnTouchOutside(false) }

    private fun remount() {
        if (!viewModel.remountState.value.isReady || !args.readOnlyFileStore!!.isReadOnly) {
            return
        }
        viewModel.remount(args.readOnlyFileStore!!)
    }

    private fun onRemountStateChanged(state: ActionState<PosixFileStore, Unit>) {
        when (state) {
            is ActionState.Ready, is ActionState.Running -> updateRemountButton()
            is ActionState.Success -> viewModel.finishRemounting()
            is ActionState.Error -> {
                val throwable = state.throwable
                throwable.printStackTrace()
                showToast(throwable.toString())
                viewModel.finishRemounting()
            }
        }
    }

    private fun updateRemountButton() {
        val textRes = when {
            viewModel.remountState.value.isRunning -> R.string.file_job_remount_loading_format
            args.readOnlyFileStore!!.isReadOnly -> R.string.file_job_remount_format
            else -> R.string.file_job_remount_success_format
        }
        binding.remountButton.text = getString(textRes, args.readOnlyFileStore!!.name())
    }

    private fun onDialogButtonClick(dialog: DialogInterface, which: Int) {
        val action = when (which) {
            DialogInterface.BUTTON_POSITIVE -> FileJobAction.POSITIVE
            DialogInterface.BUTTON_NEGATIVE -> FileJobAction.NEGATIVE
            DialogInterface.BUTTON_NEUTRAL -> FileJobAction.NEUTRAL
            else -> throw AssertionError(which)
        }
        notifyListenerOnce(action, args.showAll && binding.allCheck.isChecked)
        finish()
    }

    override fun onStart() {
        super.onStart()

        if (binding.root.parent == null) {
            val dialog = requireDialog() as AlertDialog
            val scrollView = dialog.requireViewByIdCompat<NestedScrollView>(R.id.scrollView)
            val linearLayout = scrollView.getChildAt(0) as LinearLayout
            linearLayout.addView(binding.root)
        }
    }

    override fun onCancel(dialog: DialogInterface) {
        super.onCancel(dialog)

        notifyListenerOnce(FileJobAction.CANCELED, false)
        finish()
    }

    fun onFinish() {
        notifyListenerOnce(FileJobAction.CANCELED, false)
    }

    private fun notifyListenerOnce(action: FileJobAction, isAll: Boolean) {
        if (isListenerNotified) {
            return
        }
        args.listener(action, isAll)
        isListenerNotified = true
    }

    @Parcelize
    class Args(
        val title: CharSequence,
        val message: CharSequence,
        val readOnlyFileStore: @WriteWith<ParcelableParceler> PosixFileStore?,
        val showAll: Boolean,
        val positiveButtonText: CharSequence?,
        val negativeButtonText: CharSequence?,
        val neutralButtonText: CharSequence?,
        val listener: @WriteWith<ListenerParceler>() (FileJobAction, Boolean) -> Unit
    ) : ParcelableArgs {
        object ListenerParceler : Parceler<(FileJobAction, Boolean) -> Unit> {
            override fun create(parcel: Parcel): (FileJobAction, Boolean) -> Unit =
                parcel.readParcelable<RemoteCallback>()!!.let {
                    { action, isAll ->
                        it.sendResult(Bundle().putArgs(ListenerArgs(action, isAll)))
                    }
                }

            override fun ((FileJobAction, Boolean) -> Unit).write(parcel: Parcel, flags: Int) {
                parcel.writeParcelable(RemoteCallback {
                    val args = it.getArgs<ListenerArgs>()
                    this(args.action, args.isAll)
                }, flags)
            }

            @Parcelize
            private class ListenerArgs(
                val action: FileJobAction,
                val isAll: Boolean
            ) : ParcelableArgs
        }
    }

    @Parcelize
    private class State(val isAllChecked: Boolean) : ParcelableState
}
